lib/membrane_wav/serializer.ex

```diff
@@ -0,0 +1,112 @@
+defmodule Membrane.WAV.Serializer do
+  @moduledoc """
+  Element responsible for raw audio serialization to WAV format.
+
+  Creates WAV header (its description can be found with `Membrane.WAV.Parser`) from received caps
+  and puts it before audio samples. The element assumes that audio is in PCM format. `File length`
+  and `data length` can be calculated only after processing all samples, so these values are
+  invalid (always set to 0).
+
+  The element has one option - `frames_per_buffer`. User can specify number of frames sent in one
+  buffer when demand unit on the output is `:buffers`. One frame contains `bits per sample` x
+  `number of channels` bits.
+  """
+  use Membrane.Filter
+
+  alias Membrane.Buffer
+  alias Membrane.Caps.Audio.Raw, as: Caps
+  alias Membrane.Caps.Audio.Raw.Format
+
+  @file_length 0
+  @data_length 0
+
+  @audio_format 1
+  @format_chunk_length 16
+
+  def_options frames_per_buffer: [
+                type: :integer,
+                spec: pos_integer(),
+                description: """
+                Assumed number of raw audio frames in each buffer.
+                Used when converting demand from buffers into bytes.
+                """,
+                default: 2048
+              ]
+
+  def_output_pad :output,
+    mode: :pull,
+    availability: :always,
+    caps: Caps
+
+  def_input_pad :input,
+    mode: :pull,
+    availability: :always,
+    demand_unit: :bytes,
+    caps: Caps
+
+  @impl true
+  def handle_init(state) do
+    state = Map.put(state, :header_created, false)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_caps(:input, caps, _context, state) do
+    buffer = %Buffer{payload: create_header(caps)}
+    state = Map.merge(state, %{header_created: true, caps: caps})
```

Comment:
You don't have to store caps in state, after sending them they will be available as `context.pads.output.caps`